ocpp/messages.py

```diff
@@ -360,43 +319,134 @@ class CallError:
+            self.error_code,
+        )
 
     def __repr__(self):
-        return f"<CallError - unique_id={self.unique_id}, " \
-               f"error_code={self.error_code}, " \
-               f"error_description={self.error_description}, " \
-               f"error_details={self.error_details}>"
+        return (
+            f"<CallError - unique_id={self.unique_id}, "
+            f"error_code={self.error_code}, "
+            f"error_description={self.error_description}, "
+            f"error_details={self.error_details}>"
+        )
+
+
+def validate_payload(message: Union[Call, CallResult],
+                     ocpp_version: str) -> None:
+    """Validate the payload of the message using JSON schemas."""
+    if type(message) not in [Call, CallResult]:
+        raise ValidationError(
+            "Payload can't be validated because message "
+            f"type. It's '{type(message)}', but it should "
+            "be either 'Call'  or 'CallResult'."
+        )
+
+    try:
+        # 3 OCPP 1.6 schedules have fields of type floats. The JSON schema
+        # defines a certain precision for these fields of 1 decimal. A value of
+        # 21.4 is valid, whereas a value if 4.11 is not.
+        #
+        # The problem is that Python's internal representation of 21.4 might
+        # have more than 1 decimal. It might be 21.399999999999995. This would
+        # make the validation fail, although the payload is correct. This is a
+        # known issue with jsonschemas, see:
+        # https://github.com/Julian/jsonschema/issues/247
+        #
+        # This issue can be fixed by using a different parser for floats than
+        # the default one that is used.
+        #
+        # Both the schema and the payload must be parsed using the different
+        # parser for floats.
+        if ocpp_version == "1.6" and (
+            (
+                isinstance(message, Call)
+                and message.action in ["SetChargingProfile",
+                                       "RemoteStartTransaction"]
+            )  # noqa
+            or (
+                isinstance(message, CallResult)
+                and message.action == "GetCompositeSchedule"
+            )
+        ):
+            validator = get_validator(
+                message.message_type_id,
+                message.action,
+                ocpp_version,
+                parse_float=decimal.Decimal,
+            )
+
+            message.payload = json.loads(
+                json.dumps(message.payload), parse_float=decimal.Decimal
+            )
+        else:
+            validator = get_validator(
+                message.message_type_id, message.action, ocpp_version
+            )
+    except (OSError, json.JSONDecodeError):
+        raise NotImplementedError(
+            details={"cause": f"Failed to validate action: {message.action}"}
+        )
+
+    try:
+        validator.validate(message.payload)
+    except SchemaValidationError as error:
```

Comment:
Most of the changes come from running black, the main changes were extending the error details with the ocpp_message in this part of the code